eventlet/hubs/asyncio.py

```diff
@@ -32,22 +24,46 @@ class Hub(hub.BaseHub):
 
     def __init__(self):
         super().__init__()
-        # Make sure asyncio thread pools use real threads:
-        if concurrent_imported:
-            concurrent.futures.thread.threading = original("threading")
-            concurrent.futures.thread.queue = original("queue")
-
-        # Make sure select/poll/epoll/kqueue are usable by asyncio:
-        import selectors
-        selectors.select = original("select")
-
-        # Make sure DNS lookups use normal blocking API (which asyncio will run
-        # in a thread):
-        import asyncio.base_events
-        asyncio.base_events.socket = original("socket")
+        # Make sure select/poll/epoll/kqueue are usable by asyncio, original
+        # socket.socketpair is used by asyncio, real thread pools are used,
+        # etc:
+        def import_with_original_dependencies():
+            # This gets imported super-early so random attributes point at
+            # monkey-patched select, let's try to fix that.
+            import selectors as old_selectors
+            del sys.modules["selectors"]
+            import selectors
+            # Fix old module for those who have it. Doesn't fix "from selectors
+            # import whatevs" though...
+            for attr, value in selectors.__dict__.items():
+                setattr(old_selectors, attr, value)
+
+            import asyncio
+            try:
+                import concurrent.futures.thread
+            except RuntimeError:
+                # This happens in weird edge cases where asyncio hub is started at
+                # shutdown. Not much we can do if this happens.
+                pass
+            else:
+                concurrent.futures.thread.threading = original("threading")
+                concurrent.futures.thread.queue = original("queue")
+            import asyncio.base_events
+            import asyncio.selector_events
+            asyncio.base_events.socket = original("socket")
+
+        _original_patch_function(
+            import_with_original_dependencies,
+            "select",
+            "socket",
+            "os",
+            "threading",
+            "queue"
+        )()
 
         # The presumption is that eventlet is driving the event loop, so we
         # want a new one we control.
+        import asyncio
```

Comment:
Cannot we import asyncio at the global level rather than importing it multiple time in this module?

Reply:
My understanding (if it is correct) is that we have to tweak some stdlib module like you did in this patch, but I wonder if the import of asyncio is by itself required multiple times

Reply:
The important thing is not importing it at the top level. I'll add some docs.

Reply:
ok, thanks